src/modules/firewall/src/lib/Firewall.h

```diff
@@ -39,32 +40,265 @@ class FirewallLog
+    virtual ~StringEnum() = default;
+
+    virtual bool IsValid() const = 0;
+
+    std::string ToString() const
+    {
+        return m_value;
+    }
+
+    bool operator==(const char* other) const
+    {
+        return m_value.compare(other) == 0;
+    }
+
+    bool operator!=(const char* other) const
+    {
+        return m_value.compare(other) != 0;
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const StringEnum& se)
+    {
+        os << se.ToString();
+        return os;
+    }
+
+protected:
+    std::string m_value;
+};
+
+class GenericRule
+{
+public:
+    class State : public StringEnum
+    {
+    public:
+        static const std::array<std::string, 2> m_values;
+
+        State() = default;
+        State(const std::string& value) : StringEnum(value) {}
+
+        bool IsValid() const override
+        {
+            return std::find(m_values.begin(), m_values.end(), m_value) != m_values.end();
+        }
+    };
+
+    class Action : public StringEnum
+    {
+    public:
+        static const std::array<std::string, 3> m_values;
+
+        Action() = default;
+        Action(const std::string& value) : StringEnum(value) {}
+        virtual ~Action() = default;
+
+        virtual bool IsValid() const override
+        {
+            return std::find(m_values.begin(), m_values.end(), m_value) != m_values.end();
+        }
+    };
+
+    class Direction : public StringEnum
+    {
+    public:
+        static const std::array<std::string, 2> m_values;
+
+        Direction() = default;
+        Direction(const std::string& value) : StringEnum(value) {}
+
+        bool IsValid() const override
+        {
+            return std::find(m_values.begin(), m_values.end(), m_value) != m_values.end();
+        }
+    };
+
+    class Protocol : public StringEnum
+    {
+    public:
+        static const std::array<std::string, 4> m_values;
+
+        Protocol() = default;
+        Protocol(const std::string& value) : StringEnum(value) {}
+
+        bool IsValid() const override
+        {
+            return std::find(m_values.begin(), m_values.end(), m_value) != m_values.end();
+        }
+    };
+
+    virtual GenericRule& Parse(const rapidjson::Value& rule);
+
+    virtual std::vector<std::string> GetParseError() const
+    {
+        return m_parseError;
+    }
+
+    virtual bool HasParseError() const
+    {
+        return !m_parseError.empty();
+    }
+
+    virtual State GetDesiredState() const
+    {
+        return m_desiredState;
+    }
+
+    virtual std::string Specification() const = 0;
+
+protected:
+    Action m_action;
+    Direction m_direction;
+    Protocol m_protocol;
+    std::string m_sourceAddress;
+    std::string m_destinationAddress;
+    std::string m_sourcePort;
+    std::string m_destinationPort;
+
+    std::vector<std::string> m_parseError;
+
+private:
+    State m_desiredState;
+};
+
+class GenericPolicy
+{
+public:
+    class Action : public StringEnum
+    {
+    public:
```

Comment:
Is it necessary class definition within a class definition? This makes the code more complicated than it is. Untangle and befriend, or ...?

Reply:
Agreed, move out of class definition to simplify 👍🏻